contracts/test/v0.8/PermissionedForwardProxy.ts

```diff
@@ -0,0 +1,168 @@
+
+    describe('when called by the owner', () => {
+      beforeEach(async () => {
+        // Reset permission before testing
+        await controller
+          .connect(personas.Carol)
+          .removePermission(await personas.Carol.getAddress())
+      })
+
+      it('adds the permission to the proxy', async () => {
+        await controller
+          .connect(personas.Carol)
+          .addPermission(
+            await personas.Carol.getAddress(),
+            await personas.Eddy.getAddress(),
+          )
+
+        await expect(
+          await controller.forwardPermissionList(
+            await personas.Carol.getAddress(),
+          ),
+        ).to.be.equal(await personas.Eddy.getAddress())
+      })
+    })
+  })
+
+  describe('#removePermission', () => {
+    describe('when called by a non-owner', () => {
+      it('reverts', async () => {
+        await expect(
+          controller
+            .connect(personas.Eddy)
+            .removePermission(await personas.Carol.getAddress()),
+        ).to.be.revertedWith('Only callable by owner')
+      })
+    })
+
+    describe('when called by the owner', () => {
+      beforeEach(async () => {
+        // Add permission before testing
+        await controller
+          .connect(personas.Carol)
+          .addPermission(
+            await personas.Carol.getAddress(),
+            await personas.Eddy.getAddress(),
+          )
+      })
+
+      it('removes the permission to the proxy', async () => {
+        await controller
+          .connect(personas.Carol)
+          .removePermission(await personas.Carol.getAddress())
+
+        await expect(
+          await controller.forwardPermissionList(
+            await personas.Carol.getAddress(),
+          ),
+        ).to.be.equal('0x0000000000000000000000000000000000000000') // Defaults to 0x0 address
+      })
+    })
+  })
+
+  describe('#forward', () => {
+    describe('when permission does not exist', () => {
+      beforeEach(async () => {
+        // Reset permission before testing
+        await controller
+          .connect(personas.Carol)
+          .removePermission(await personas.Carol.getAddress())
+      })
+      it('reverts', async () => {
+        await expect(
+          controller
+            .connect(personas.Carol)
+            .forward(await personas.Eddy.getAddress(), '0x'),
+        ).to.be.revertedWith('Forwarding permission not found')
+      })
+    })
+
+    describe('when permission exists', () => {
+      let counter: Contract
+      beforeEach(async () => {
+        // Deploy Counter contract to call
+        counter = await counterFactory.connect(personas.Carol).deploy()
+
+        // Add permission before testing
+        await controller
+          .connect(personas.Carol)
+          .addPermission(await personas.Carol.getAddress(), counter.address)
+      })
+
+      it('calls target successfully', async () => {
+        let encoder = new ethers.utils.Interface(['function increment()'])
```

Comment:
this totally works but wanted to point out you can use the contract itself to encode these, and that way you get type checking on the function name and args. Should be like `counter.interface.encodeFunctionData()`.